<a href="https://colab.research.google.com/github/llrt/alura-imersaodados3/blob/main/alura_imersaodados3_aula1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Imersão Dados da Alura - Aula 1

Análise exploratória de dados em cima de problema de _drug discovery_

## Importação de bibliotecas e leitura de base dados

In [2]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all" # config para printar todos os statements, nao apenas o último

import pandas as pd 

pd.set_option('display.max_rows', 300) # setar nr de linhas mostradas
pd.set_option('display.max_columns', 50) # setar nr de colunas mostradas

In [3]:
# Ler base de dados de interesse
dados = pd.read_csv("https://github.com/llrt/alura-imersaodados3/blob/main/dados/dados_experimentos.zip?raw=true", 
            compression="zip") # como o arquivo é um CSV zipado, utilizar opção para tratamento automático da compressão

## Dados gerais da base de dados


In [4]:
dados.shape # dimensões da base de dados (nr de linhas, nr de colunas)

(23814, 877)

In [5]:
dados.info() # dados gerais das colunas

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23814 entries, 0 to 23813
Columns: 877 entries, id to c-99
dtypes: float64(872), int64(1), object(4)
memory usage: 159.3+ MB


In [6]:
dados.head(n=6) # 6 primeiras linhas da base
dados.tail(n=6) # 6 últimas linhas da base

id tratamento  tempo dose      droga     g-0     g-1     g-2  \
0  id_000644bb2  com_droga     24   D1  b68db1d53  1.0620  0.5577 -0.2479   
1  id_000779bfc  com_droga     72   D1  df89a8e5a  0.0743  0.4087  0.2991   
2  id_000a6266a  com_droga     48   D1  18bb41b2c  0.6280  0.5817  1.5540   
3  id_0015fd391  com_droga     48   D1  8c7f86626 -0.5138 -0.2491 -0.2656   
4  id_001626bd3  com_droga     72   D2  7cbed3131 -0.3254 -0.4009  0.9700   
5  id_001762a82  com_droga     24   D1  e06749542 -0.6111  0.2941 -0.9901   

      g-3     g-4     g-5     g-6     g-7     g-8     g-9    g-10    g-11  \
0 -0.6208 -0.1944 -1.0120 -1.0220 -0.0326  0.5548 -0.0921  1.1830  0.1530   
1  0.0604  1.0190  0.5207  0.2341  0.3372 -0.4047  0.8507 -1.1520 -0.4201   
2 -0.0764 -0.0323  1.2390  0.1715  0.2155  0.0065  1.2300 -0.4797 -0.5631   
3  0.5288  4.0620 -0.8095 -1.9590  0.1792 -0.1321 -1.0600 -0.8269 -0.3584   
4  0.6919  1.4180 -0.8244 -0.2800 -0.1498 -0.8789  0.8630 -0.2219 -0.5121   
5  0.2277  1.2810  0.5203  0.0543 -0.2225 -0.1586  0.4289  0.0361  0.3543   

     g-12    g-13    g-14    g-15    g-16    g-17    g-18    g-19  ...  \
0  0.5574 -0.4015  0.1789 -0.6528 -0.7969  0.6342  0.1778 -0.3694  ...   
1 -0.0958  0.4590  0.0803  0.2250  0.5293  0.2839 -0.3494  0.2883  ...   
2 -0.0366 -1.8300  0.6057 -0.3278  0.6042 -0.3075 -0.1147 -0.0570  ...   
3 -0.8511 -0.5844 -2.5690  0.8183 -0.0532 -0.8554  0.1160 -2.3520  ...   
4 -0.9577  1.1750  0.2042  0.1970  0.1244 -1.7090 -0.3543 -0.5160  ...   
5  0.8310 -0.9129  0.7677 -0.2512  0.7087  0.0481 -0.4486  1.2570  ...   

     c-75    c-76    c-77    c-78    c-79    c-80    c-81    c-82    c-83  \
0  0.0495  0.4141  0.8432  0.6162 -0.7318  1.2120  0.6362 -0.4427  0.1288   
1  0.7639  0.5499 -0.3322 -0.0977  0.4329 -0.2782  0.7827  0.5934  0.3402   
2  0.9760 -0.0427 -0.1235  0.0959  0.0690 -0.9416 -0.7548 -0.1109 -0.6272   
3 -1.0810 -3.0520 -3.4470 -2.7740 -1.8460 -0.5568 -3.3960 -2.9510 -1.1550   
4  0.4817  0.0587  0.5303  0.6376 -0.3966 -1.4950 -0.9625 -0.0541  0.6273   
5  0.2152  1.6920 -0.2930  1.4590  1.0950  0.5854  1.7270  0.5580  1.3090   

     c-84    c-85    c-86    c-87    c-88    c-89    c-90    c-91    c-92  \
0  1.4840  0.1799  0.5367 -0.1111 -1.0120  0.6685  0.2862  0.2584  0.8076   
1  0.1499  0.4420  0.9366  0.8193 -0.4236  0.3192 -0.4265  0.7543  0.4708   
2  0.3019  0.1172  0.1093 -0.3113  0.3019 -0.0873 -0.7250 -0.6297  0.6103   
3 -3.2620 -1.5390 -2.4600 -0.9417 -1.5550  0.2431 -2.0990 -0.6441 -5.6300   
4  0.4563  0.0698  0.8134  0.1924  0.6054 -0.1824  0.0042  0.0048  0.6670   
5  1.0100  0.6980  0.5455  1.0170  1.2910  1.5430  1.8390  1.1570 -1.0120   

     c-93    c-94    c-95    c-96    c-97    c-98    c-99  
0  0.5523 -0.1912  0.6584 -0.3981  0.2139  0.3801  0.4176  
1  0.0230  0.2957  0.4899  0.1522  0.1241  0.6077  0.7371  
2  0.0223 -1.3240 -0.3174 -0.6417 -0.2187 -1.4080  0.6931  
3 -1.3780 -0.8632 -1.2880 -1.6210 -0.8784 -0.3876 -0.8154  
4  1.0690  0.5523 -0.3031  0.1094  0.2885 -0.3786  0.7125  
5  1.9010  1.4270  0.4519  1.2120  0.3765  0.7848  1.3990  

[6 rows x 877 columns]

id    tratamento  tempo dose      droga     g-0     g-1  \
23808  id_fff8c2444     com_droga     72   D1  6c3a459be  0.1608 -1.0500   
23809  id_fffb1ceed     com_droga     24   D2  df1d0a5a1  0.1394 -0.0636   
23810  id_fffb70c0c     com_droga     24   D2  ecf3b6b74 -1.3260  0.3478   
23811  id_fffc1c3f4  com_controle     48   D2  cacb2b860  0.3942  0.3756   
23812  id_fffcb9e7c     com_droga     24   D1  8b87a7a83  0.6660  0.2324   
23813  id_ffffdd77b     com_droga     72   D1  972f41291 -0.8598  1.0240   

          g-2     g-3     g-4     g-5     g-6     g-7     g-8     g-9    g-10  \
23808  0.2551 -0.2239 -0.2431  0.4256 -0.1166 -0.1777 -0.7480  0.1368  0.5493   
23809 -0.1112 -0.5080 -0.4713  0.7201  0.5773  0.3055 -0.4726  0.1269  0.2531   
23810 -0.3743  0.9905 -0.7178  0.6621 -0.2252 -0.5565  0.5112  0.6727 -0.1851   
23811  0.3109 -0.7389  0.5505 -0.0159 -0.2541  0.1745 -0.0340  0.4865 -0.1854   
23812  0.4392  0.2044  0.8531 -0.0343  0.0323  0.0463  0.4299 -0.7985  0.5742   
23813 -0.1361  0.7952 -0.3611 -3.6750 -1.2420  0.9146  3.0790  1.2460  1.9460   

         g-11    g-12    g-13    g-14    g-15    g-16    g-17    g-18    g-19  \
23808 -0.5083  0.4112  0.1181  0.7861 -0.1877  0.2476  0.1956  0.0273  0.1182   
23809  0.1730 -0.4532 -1.0790  0.2474 -0.4550  0.3588  0.1600 -0.7362 -0.1103   
23810  2.8650 -0.2140 -0.6153  0.8362  0.5584 -0.2589  0.1292  0.0148  0.0949   
23811  0.0716  0.1729 -0.0434  0.1542 -0.2192 -0.0302 -0.4218  0.4057 -0.5372   
23812  0.1421  2.2700  0.2046  0.5363 -1.7330  0.1450  0.6097  0.2024  0.9865   
23813  1.4370  2.9780  2.2370 -0.6818  0.6870 -1.1060  0.0182 -0.9247 -0.0738   

       ...    c-75    c-76    c-77    c-78    c-79    c-80    c-81    c-82  \
23808  ...  0.6498  0.8293 -0.7645  0.5379  0.5610  0.2577 -0.5493  0.2690   
23809  ...  0.3119 -0.7697  0.2203 -1.0710  0.5979  0.0848 -0.2555  0.6293   
23810  ...  0.2408 -0.7781 -0.0929 -0.0329  0.0781 -1.4440 -0.0921  0.3188   
23811  ... -0.1882 -0.0087 -0.0356  0.5718  0.4971  0.1057  0.2111  0.6992   
23812  ...  1.0340 -0.0270 -0.4194  0.7403 -0.6793 -0.1423  0.7307  0.7946   
23813  ... -2.2490 -2.5860 -1.9520 -2.1810 -4.6690 -3.9450 -2.9560 -2.7930   

         c-83    c-84    c-85    c-86    c-87    c-88    c-89    c-90    c-91  \
23808  0.6392 -0.5774 -0.3393  0.5068 -0.2005 -0.4559  0.3734  0.0789  0.3538   
23809  1.1660  0.3329  0.2754  0.4108 -0.1252 -0.2340  0.2267  0.1969  0.0262   
23810 -1.1080  0.4895 -0.2144  1.0960  0.3335  0.4444 -1.1130  0.4286  0.4426   
23811  0.0708  0.6169  0.2248  0.5994  0.2689  0.0305  1.2320  0.5409  0.3755   
23812 -0.0650  0.9038  0.2324  0.9676  1.0940  0.4290  0.5187 -0.1105  0.4258   
23813 -2.1560 -2.4100 -1.8190 -3.3480 -0.1414 -2.6430 -2.5810 -3.3890 -1.7450   

         c-92    c-93    c-94    c-95    c-96    c-97    c-98    c-99  
23808  0.0558  0.3377 -0.4753 -0.2504 -0.7415  0.8413 -0.4259  0.2434  
23809 -0.8121  0.3434  0.5372 -0.3246  0.0631  0.9171  0.5258  0.4680  
23810  0.0423 -0.3195 -0.8086 -0.9798 -0.2084 -0.1224 -0.2715  0.3689  
23811  0.7343  0.2807  0.4116  0.6422  0.2256  0.7592  0.6656  0.3808  
23812 -0.2012  0.1506  1.5230  0.7101  0.1732  0.7015 -0.6290  0.0740  
23813 -6.6300 -4.0950 -7.3860 -1.4160 -3.5770 -0.4775 -2.1500 -4.2520  

[6 rows x 877 columns]

In [7]:
# ou simplesmente
dados

id    tratamento  tempo dose      droga     g-0     g-1  \
0      id_000644bb2     com_droga     24   D1  b68db1d53  1.0620  0.5577   
1      id_000779bfc     com_droga     72   D1  df89a8e5a  0.0743  0.4087   
2      id_000a6266a     com_droga     48   D1  18bb41b2c  0.6280  0.5817   
3      id_0015fd391     com_droga     48   D1  8c7f86626 -0.5138 -0.2491   
4      id_001626bd3     com_droga     72   D2  7cbed3131 -0.3254 -0.4009   
...             ...           ...    ...  ...        ...     ...     ...   
23809  id_fffb1ceed     com_droga     24   D2  df1d0a5a1  0.1394 -0.0636   
23810  id_fffb70c0c     com_droga     24   D2  ecf3b6b74 -1.3260  0.3478   
23811  id_fffc1c3f4  com_controle     48   D2  cacb2b860  0.3942  0.3756   
23812  id_fffcb9e7c     com_droga     24   D1  8b87a7a83  0.6660  0.2324   
23813  id_ffffdd77b     com_droga     72   D1  972f41291 -0.8598  1.0240   

          g-2     g-3     g-4     g-5     g-6     g-7     g-8     g-9    g-10  \
0     -0.2479 -0.6208 -0.1944 -1.0120 -1.0220 -0.0326  0.5548 -0.0921  1.1830   
1      0.2991  0.0604  1.0190  0.5207  0.2341  0.3372 -0.4047  0.8507 -1.1520   
2      1.5540 -0.0764 -0.0323  1.2390  0.1715  0.2155  0.0065  1.2300 -0.4797   
3     -0.2656  0.5288  4.0620 -0.8095 -1.9590  0.1792 -0.1321 -1.0600 -0.8269   
4      0.9700  0.6919  1.4180 -0.8244 -0.2800 -0.1498 -0.8789  0.8630 -0.2219   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
23809 -0.1112 -0.5080 -0.4713  0.7201  0.5773  0.3055 -0.4726  0.1269  0.2531   
23810 -0.3743  0.9905 -0.7178  0.6621 -0.2252 -0.5565  0.5112  0.6727 -0.1851   
23811  0.3109 -0.7389  0.5505 -0.0159 -0.2541  0.1745 -0.0340  0.4865 -0.1854   
23812  0.4392  0.2044  0.8531 -0.0343  0.0323  0.0463  0.4299 -0.7985  0.5742   
23813 -0.1361  0.7952 -0.3611 -3.6750 -1.2420  0.9146  3.0790  1.2460  1.9460   

         g-11    g-12    g-13    g-14    g-15    g-16    g-17    g-18    g-19  \
0      0.1530  0.5574 -0.4015  0.1789 -0.6528 -0.7969  0.6342  0.1778 -0.3694   
1     -0.4201 -0.0958  0.4590  0.0803  0.2250  0.5293  0.2839 -0.3494  0.2883   
2     -0.5631 -0.0366 -1.8300  0.6057 -0.3278  0.6042 -0.3075 -0.1147 -0.0570   
3     -0.3584 -0.8511 -0.5844 -2.5690  0.8183 -0.0532 -0.8554  0.1160 -2.3520   
4     -0.5121 -0.9577  1.1750  0.2042  0.1970  0.1244 -1.7090 -0.3543 -0.5160   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
23809  0.1730 -0.4532 -1.0790  0.2474 -0.4550  0.3588  0.1600 -0.7362 -0.1103   
23810  2.8650 -0.2140 -0.6153  0.8362  0.5584 -0.2589  0.1292  0.0148  0.0949   
23811  0.0716  0.1729 -0.0434  0.1542 -0.2192 -0.0302 -0.4218  0.4057 -0.5372   
23812  0.1421  2.2700  0.2046  0.5363 -1.7330  0.1450  0.6097  0.2024  0.9865   
23813  1.4370  2.9780  2.2370 -0.6818  0.6870 -1.1060  0.0182 -0.9247 -0.0738   

       ...    c-75    c-76    c-77    c-78    c-79    c-80    c-81    c-82  \
0      ...  0.0495  0.4141  0.8432  0.6162 -0.7318  1.2120  0.6362 -0.4427   
1      ...  0.7639  0.5499 -0.3322 -0.0977  0.4329 -0.2782  0.7827  0.5934   
2      ...  0.9760 -0.0427 -0.1235  0.0959  0.0690 -0.9416 -0.7548 -0.1109   
3      ... -1.0810 -3.0520 -3.4470 -2.7740 -1.8460 -0.5568 -3.3960 -2.9510   
4      ...  0.4817  0.0587  0.5303  0.6376 -0.3966 -1.4950 -0.9625 -0.0541   
...    ...     ...     ...     ...     ...     ...     ...     ...     ...   
23809  ...  0.3119 -0.7697  0.2203 -1.0710  0.5979  0.0848 -0.2555  0.6293   
23810  ...  0.2408 -0.7781 -0.0929 -0.0329  0.0781 -1.4440 -0.0921  0.3188   
23811  ... -0.1882 -0.0087 -0.0356  0.5718  0.4971  0.1057  0.2111  0.6992   
23812  ...  1.0340 -0.0270 -0.4194  0.7403 -0.6793 -0.1423  0.7307  0.7946   
23813  ... -2.2490 -2.5860 -1.9520 -2.1810 -4.6690 -3.9450 -2.9560 -2.7930   

         c-83    c-84    c-85    c-86    c-87    c-88    c-89    c-90    c-91  \
0      0.1288  1.4840  0.1799  0.5367 -0.1111 -1.0120  0.6685  0.2862  0.2584   
1      0.3402  0.1499  0.4420  0.9366  0.8193 -0.4236  0.31

## Análise univariada (i.e. de cada coluna isolada)

### Coluna _tratamento_

Quais são os valores únicos da coluna tratamento? Qual a sua semântica?

In [43]:
dados['tratamento'].value_counts() # quais valores únicos existem na coluna?

com_droga       21948
com_controle     1866
Name: tratamento, dtype: int64

In [42]:
dados['tratamento'].value_counts(normalize=True) # qual a proporção entre os valores únicos encontrados

com_droga       0.921643
com_controle    0.078357
Name: tratamento, dtype: float64

A coluna _tratamento_ indica se o experimento em questão se refere a um grupo de controle (_com_controle_ ) ou de teste (_com_droga_). Note que a proporcao é muito desbalanceada (92% com droga). Por quê?

com_droga       0.921643
com_controle    0.078357
Name: tratamento, dtype: float64

### Coluna _droga_

A coluna droga identifica de maneira única a droga/fármaco (ou controle) testado naquele experimento. Quantas são as drogas testadas?

In [44]:
dados['droga'].value_counts()

cacb2b860    1866
87d714366     718
9f80f3f77     246
8b87a7a83     203
5628cb3ee     202
             ... 
c5ffd58ca       1
0e6ee26ff       1
573c787a2       1
4055d26a8       1
82dcdb180       1
Name: droga, Length: 3289, dtype: int64

Parecem muitas drogas! Na verdade não são fármacos finais, mas são compostos que estão sendo testados.

## Conclusões

* A base registra diversos experimentos com fármacos, num total de 23813 experimentos, com um total de 877 colunas de informações
* Além de colunas de id do experimento (campo *id*), temos colunas que informam: 
** Se se trata de experimento do grupo de controle (sem uso de fármaco ou placebo) ou com uso de fármaco - campo *tratamento*
** Qual o tempo de uso do fármaco - campo *tempo*
** O tipo de dose aplicada (D1 ou D2, a investigar o que significam) - campo *dose*
** O id da droga investigada - campo *droga*
** A expressão gênica naquele experimento - campos *g-\**

## Dúvidas

* O que significam as doses D1 e D2? São doses de tamanhos diferentes?
* Os campos *g-\** estão com dados normalizados?
* O que são os campos *c-\** ?